embassy/src/util/mpsc.rs

```diff
@@ -0,0 +1,905 @@
+///
+/// Note that when splitting the channel, the sender and receiver cannot outlive
+/// their channel. The following will therefore fail compilation:
+////
+/// ```compile_fail
+/// use embassy::util::mpsc;
+/// use embassy::util::mpsc::{Channel, WithThreadModeOnly};
+///
+/// let (sender, receiver) = {
+///    let mut channel = Channel::<WithThreadModeOnly, u32, 3>::with_thread_mode_only();
+///     mpsc::split(&channel)
+/// };
+/// ```
+pub fn split<M, T, const N: usize>(
+    channel: &Channel<M, T, N>,
+) -> (Sender<M, T, N>, Receiver<M, T, N>)
+where
+    M: Mutex<Data = ()>,
+{
+    let sender = Sender { channel: &channel };
+    let receiver = Receiver { channel: &channel };
+    {
+        let c = channel.get();
+        c.register_receiver();
+        c.register_sender();
+    }
+    (sender, receiver)
+}
+
+impl<'ch, M, T, const N: usize> Receiver<'ch, M, T, N>
+where
+    M: Mutex<Data = ()>,
+{
+    /// Receives the next value for this receiver.
+    ///
+    /// This method returns `None` if the channel has been closed and there are
+    /// no remaining messages in the channel's buffer. This indicates that no
+    /// further values can ever be received from this `Receiver`. The channel is
+    /// closed when all senders have been dropped, or when [`close`] is called.
+    ///
+    /// If there are no messages in the channel's buffer, but the channel has
+    /// not yet been closed, this method will sleep until a message is sent or
+    /// the channel is closed.
+    ///
+    /// Note that if [`close`] is called, but there are still outstanding
+    /// messages from before it was closed, the channel is not considered
+    /// closed by `recv` until they are all consumed.
+    ///
+    /// [`close`]: Self::close
+    pub async fn recv(&mut self) -> Option<T> {
+        self.await
+    }
+
+    /// Attempts to immediately receive a message on this `Receiver`
+    ///
+    /// This method will either receive a message from the channel immediately or return an error
+    /// if the channel is empty.
+    pub fn try_recv(&self) -> Result<T, TryRecvError> {
+        self.channel.get().try_recv()
+    }
+
+    /// Closes the receiving half of a channel without dropping it.
+    ///
+    /// This prevents any further messages from being sent on the channel while
+    /// still enabling the receiver to drain messages that are buffered.
+    ///
+    /// To guarantee that no messages are dropped, after calling `close()`,
+    /// `recv()` must be called until `None` is returned. If there are
+    /// outstanding messages, the `recv` method will not return `None`
+    /// until those are released.
+    ///
+    pub fn close(&mut self) {
+        self.channel.get().close()
+    }
+}
+
+impl<'ch, M, T, const N: usize> Future for Receiver<'ch, M, T, N>
+where
+    M: Mutex<Data = ()>,
+{
+    type Output = Option<T>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        match self.try_recv() {
```

Comment:
This can lead to a race condition. `try_recv` and `set_receiver_waker` both lock the channel state independently, in between them it is unlocked and another thread could affect the state. This is a problem, for example:

- This thread calls try_recv, it returns Empty
- Another thread pushes a value. No waker is registered yet, so it doesn't wake anything.
- This thread calls `set_receiver_waker`, registering the waker even though the channel is no longer empty.
- End result: current task doesn't pop the value out, and waits forever instead.

Easiest way to avoid this is to ensure each "operation" such as this poll only locks the mutex once, so that it's atomic.

Reply:
Well spotted! Fixed as per your suggestion of an atomic operation in commit 196f555e61571055fd8e111bbbf3358c895fe01c